keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -414,48 +418,73 @@ Both these cases will be handled properly, here's an example:
 type Object struct {
      // This field has an invalid zero-value, empty string is not a
      // valid string.
-     // +default=default-name
-     Name string
+     //
+     // If this is a CRD, there would be a kubebuilder specific tag, and
+     // if this is a built-in, there would be a specific validation function,
+     // but basically:
+     // +minLength=1
+     //
+     // +default="default-name"
+     Name string `json:",omitempty"`
 
      // This field is defaulted to 0 with no defaulter.
      // +default=0
      Defaulted int
 }
 ```
 
+Non-pointer fields that have a default must be omitEmpty. Kubebuilder
+and built-in defaulting generators should both prevent these from
+happening.
+
 This would generate the following OpenAPI:
 ```yaml
 Object:
   type: object
   properties:
     invalid:
       type: string
-      default: default-name
+      default: "default-name"
+      # This field would only be present for CRDs, for built-in,
+      # the empty value would be accepted and treated as missing.
+      minLength: 1
     defaulted:
-      type: boolean
+      type: int
       default: false
 ```
 
 Which, for both CRD types and built-in types, would generate the following behavior:
 ```python
 >>> default('{}')
 {"entry": {"name": "default-name", "number": 0}}
->>> default('{entry: {name: other-name}}')
-{"entry": {"name": "other-name", "number": 0}}
+>>> default('{name: other-name}')
+{"name": "other-name", "number": 0}
+# For built-in, additionally, we would have:
+>>> default('{name: ""}')
+{"name": "default-name", "number": 0}
 ```
 
-Note that the generated code for built-in will look like this:
+Note that a simplified, and non-cached version of the generated code for
+built-ins would look like this:
 ```golang
 func GenerateDefault(obj *Object) {
      if obj.Name == "" {
-          obj.Name = "default-name"
+          if err := json.Unmarshal([]byte(`"default-name"`), &obj.Name); err != nil {
+               panic(err)
+          }
      }
      if obj.Number == 0 {
-          obj.Number = 0
+          if err := json.Unmarshal([]byte(`0`), &obj.Number); err != nil {
+               panic(err)
+          }
```

Comment:
I hope this is just a sketch and we are a bit more clever than calling Unmarshal all the time.

Reply:
Yes, we're clearly planning on caching the deserialized version since it doesn't change.